components/layout_2020/flow/inline/mod.rs

```diff
@@ -727,15 +728,27 @@ impl<'layout_dta> InlineFormattingContextLayout<'layout_dta> {
             );
         }
 
+        // Boxes Padding - Border - Margin (PBM) and positioning context should follow
+        // visual, not logical order.
+        // https://drafts.csswg.org/css-writing-modes-4/#bidi-fragment-boxes
+        // 2.4.5.2. Box Model of Reordering-induced Box Fragments
+        let css_direction: ComputedCSSDirection = inline_box_state.base.style.clone_direction();
+
         if inline_box.is_first_fragment {
             self.current_line_segment.inline_size += inline_box_state.pbm.padding.inline_start +
                 inline_box_state.pbm.border.inline_start +
                 inline_box_state.pbm.margin.inline_start.auto_is(Au::zero);
             self.current_line_segment
                 .line_items
-                .push(LineItem::LeftInlineBoxPaddingBorderMargin(
-                    inline_box.identifier,
-                ));
+                .push(match css_direction {
+                        ComputedCSSDirection::Ltr => {
+                            LineItem::LeftInlineBoxPaddingBorderMargin(inline_box.identifier)
```

Comment:
I think these should be renamed to `InlineStartBoxPaddingBorderMargin` and `InlineEndBoxPaddingBorderMargin` or such, then you wouldn't need to check the direction here.

Instead, https://github.com/servo/servo/blob/bd37d5496d747c7106b2d8d73dca1212421d82b4/components/layout_2020/flow/inline/line.rs#L379-L389 should just swap if the direction doesn't match the containing block.

Reply:
I don't think that it is proper change. Because if we change order of items in inline_items container. Then inline boxes borders could potentially participate in general procedure of text runs and inline items bidi reordering.
I believe that we will need it in the future to pass this test.
![image](https://github.com/user-attachments/assets/4cbc425e-f4ef-46c7-8e64-17c45af563b1)

Reply:
Currently there are multiple problems why this test fail. Issues with white space collapsing. Issues with inline boxes borders reordering. I need to investigate but my current idea is that borders should also participate in reordering procedure.

Reply:
So can we have some talk regarding this issue? 
I will need to prepare for such talk. Will reread documentation once again and make some notes.

I understand that in the Draft authors explicitly talk about fragments. And initially I have idea of implementing it near the place that you suggest. TBH I haven't find exactly the place that you stated in comment.

So we just need to understand will boxes ever participate in reordering process together with text_runs

Reply:
@Loirooriol

Reply:
![image](https://github.com/user-attachments/assets/17df768e-f813-423b-83ba-9af8ec0aa449)

Reply:
So currently there is some confusion in specs I believe. It would be cool if CSS committee can decide where should we apply bidi reordering and clearly state that. Theoretically it could be done on fragments level purely. Or we need more clearly define 2 separate procedures.
I.E text_runs bidi reordering and box fragments reordering as a separate term.

Reply:
@Loirooriol Look into Zulip.
Maybe I should to the opposite and extract this from fragment level to inline_items level?

Reply:
I can experiment tomorrow by adding control symbols to each inline item. Instead of simple copy of the last bidi level. And look into general behavior

Reply:
> I don't think that it is proper change. Because if we change order of items in inline_items container

I'm confused, I didn't mean to reorder anything? I will push the change that I meant.